File: game_class.py
import copy
import numpy as np

class Game:
    def __init__(self, inherit=None) -> None:
        if not inherit:
            self.mode = "start"

            # moves amount
            self.moves_amount: int = 0

            # check
            self.check: bool = False

            # en passant
            self.en_passant: list[ int ] = [0, 0]

            # castle switches
            self.castle: dict[ str, bool ] = {'left_w': True, 'right_w': True, 'left_b': True, 'right_b': True}

            # Player dependant variables
            self.player: str = '' # [ 'w' , 'b' ]
            self.opponent: str = '' # [ 'w' , 'b' ]

            # dictionaries for all pieces and their moves
            self.moves: dict = {}

            # main board
            self.board = np.array([
                ["bk", "  ", "  ", "  ", "  ", "  ", "  ", "  "],
                ["bp", "bp", "bp", "bp", "bp", "bp", "bp", "bp"],
                ["  ", "  ", "  ", "  ", "  ", "  ", "  ", "  "],
                ["  ", "  ", "  ", "  ", "  ", "  ", "  ", "  "],
                ["wp", "  ", "  ", "  ", "wq", "  ", "  ", "  "],
                ["  ", "  ", "  ", "  ", "  ", "  ", "  ", "  "],
                ["wp", "wp", "wp", "  ", "  ", "  ", "  ", "  "],
                ["  ", "  ", "  ", "  ", "  ", "  ", "  ", "  "],
            ])

            # captured pieces
            self.captured: dict[str, list] = {'w': [], 'b': []}

        else:
            self.mode = inherit.mode
            # moves amount
            self.moves_amount: int = inherit.moves_amount
            # check
            self.check: bool = inherit.check
            # en passant
            self.en_passant: list[int] = inherit.en_passant.copy()
            # castle switches
            self.castle: dict[str, bool] = inherit.castle.copy()
            # Player dependant variables
            self.player: str = inherit.player
            self.opponent: str = inherit.opponent
            # dictionaries for all pieces and their moves
            self.moves: dict = copy.deepcopy(inherit.moves)
            self.board = inherit.board.copy()
            self.captured: dict[str, list] = copy.deepcopy(inherit.captured)

File: rust/src/state/game_class.rs
// Saves all the info about the game
#[derive(Clone)]
pub struct Game {

    // game info
    pub board: [[(char,char);8];8],
    pub en_passant: i8,
    pub castle: [[bool;2];2],
    pub check: bool,

    // kings
    pub w_king_pos: (i8,i8),
    pub b_king_pos: (i8,i8),

    // info for stoping the game
    pub mode: char,
    pub moves_amount: i16,
    
    // players
    pub player: char,
    pub opponent: char,

    // moves
    pub w_cover: [[Vec<(i8,i8)>;8];8],
    pub b_cover: [[Vec<(i8,i8)>;8];8],
    pub legal:   [[Vec<(i8,i8)>;8];8],

    // history
    pub captured: Vec<(char,char)>,
    pub history: Vec<[[(char,char);8];8]>,

}

impl Game {
    pub fn new() -> Self {
        Game {
            // starting board
            board: [
                [('b','r'), ('b','h'), ('b','b'), ('b','q'), ('b','k'), ('b','b'), ('b','h'), ('b','r')],
                [('b','p'), ('b','p'), ('b','p'), ('b','p'), ('b','p'), ('b','p'), ('b','p'), ('b','p')],
                [(' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' ')],
                [(' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' ')],
                [(' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' ')],
                [(' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' '), (' ',' ')],
                [('w','p'), ('w','p'), ('w','p'), ('w','p'), ('w','p'), ('w','p'), ('w','p'), ('w','p')],
                [('w','r'), ('w','h'), ('w','b'), ('w','q'), ('w','k'), ('w','b'), ('w','h'), ('w','r')],
            ],
            
            // game info
            en_passant: -1,
            castle: [[true,true],[true,true]],
            check: false,

            //kings
            w_king_pos: (7,4),
            b_king_pos: (0,4),

            // info for stoping the game
            mode: 'g',
            moves_amount: 0,

            // players
            player: 'w',
            opponent: 'b',

            // moves
            w_cover: std::array::from_fn(|_| std::array::from_fn(|_| Vec::new())),
            b_cover: std::array::from_fn(|_| std::array::from_fn(|_| Vec::new())),
            legal:   std::array::from_fn(|_| std::array::from_fn(|_| Vec::new())),
            
            // history
            captured: Vec::new(),
            history: Vec::new(),
        }
    }


    // get player cover
    pub fn pl_cover(&self) -> &[[Vec<(i8,i8)>;8];8] {
        if self.player == 'w' {
            return &self.w_cover;
        } else {
            return &self.b_cover;
        }
    }

    // get opponent cover
    pub fn op_cover(&self) -> &[[Vec<(i8,i8)>;8];8] {
        if self.player == 'b' {
            return &self.w_cover;
        } else {
            return &self.b_cover;
        }
    }

    pub fn king_pos(&self, side: char) -> &(i8,i8) {
        if side == 'w' {
            return &self.w_king_pos;
        } else {
            return &self.b_king_pos;
        }
    }
}



